### PR TITLE
Add Neutron l3 Agent Sensu Check

### PR DESCRIPTION
**Pull request by ddaskal**
I have modified this check to run using python-neutronclient.
I have set a maximum number of routers within the script to 100.
This is in accordance with: https://github.com/blueboxgroup/ursula/pull/2301

### CODE REVIEW

**Comment by ddaskal** [comment]
I've included both commits to show the requests method, and then the python-neutronclient method. I can squash when the latest one is accepted.

File: sensu/plugins/check-neutron-l3-routers.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                         project_name=project_name,
+                         auth_url=auth_url)
+
+sess = session.Session(auth=auth)
+neutron = neutron_client.Client(session=sess)
+
+STATE_OK = 0
+STATE_WARNING = 1
+STATE_CRITICAL = 2
+MAX_ROUTERS = 100
```
Comment (craigtracey) [inline sensu/plugins/check-neutron-l3-routers.py]: Make this configurable at the command line

File: sensu/plugins/check-neutron-l3-routers.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def check_router(router, router_agents):
+
```
Comment (craigtracey) [inline sensu/plugins/check-neutron-l3-routers.py:44]: It would be good to comment what the goal is here. This is a bit hard to follow.

File: sensu/plugins/check-neutron-l3-routers.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    active_agent = False
+    bad_agents = False
+
+    for agent in router_agents['agents']:
+        if 'ha_state' not in agent.keys():
+            print("WARNING: No HA state for l3 agent %s" % agent['id'])
```
Comment (craigtracey) [inline sensu/plugins/check-neutron-l3-routers.py:49]: Here you are issuing a warning but not bubbling up the warning.  Perhaps a better way to do this would be to return the const itself.

File: sensu/plugins/check-neutron-l3-routers.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    router_list = neutron.list_routers()
+    if len(router_list['routers']) > MAX_ROUTERS:
+        print("WARNING: Number of routers more than the max: %d" % MAX_ROUTERS)
+        sys.exit(STATE_WARNING)
+
+    for router in router_list['routers']:
+        search_opts['router'] = router['id']
+        router_agents = neutron.list_l3_agent_hosting_routers(**search_opts)
+        bad_agents = check_router(router, router_agents)
```
Comment (craigtracey) [inline sensu/plugins/check-neutron-l3-routers.py]: It would be good to capture the bad ones.  Perhaps we should do that here?

Comment (craigtracey) [reply sensu/plugins/check-neutron-l3-routers.py]: Additionally, it might be good to have a threshold for "bad agents"

**Comment by craigtracey** [comment]
Let me know if these comments are not clear...we should be able to hammer them out quickly. Ping directly if anything is not clear.

**Comment by nirajdp76** [comment]
is this check only going to be active/making neutron calls on controller with ucarp?

**Comment by ddaskal** [comment]
@nirajdp76 I have the task in Ursula set to only run on the first controller.
```when: inventory_hostname in groups['controller'][0]```

**Comment by craigtracey** [comment]
This check is fine but the Ursula logic will not work.  What happens when the controller fails over?

**Review by craigtracey — APPROVED** [review]

**Comment by ddaskal** [comment]
@craigtracey Even if we set the task to run on the current UCARP primary, only that controller will contain the json config for the check in `/etc/sensu/conf.d/checks/`. So after a post-install fail over, the check will still run from the new UCARP secondary controller whenever it reboots.
